### ISSUE TITLE
UI is blurry in 4.1.0 when system-wide scaling is set to 100%

### ISSUE THREAD

**Issue by lexi-lambda** [EDCD/EDMarketConnector#734]
- Version 4.1.0
 - OS: Windows 10
 - OS Locale: English
 - [EDMarketConnector.log](https://github.com/EDCD/EDMarketConnector/files/5331912/EDMarketConnector.log)
 - [EDMarketConnector-debug.log](https://github.com/EDCD/EDMarketConnector/files/5331914/EDMarketConnector-debug.log)

### Describe the bug

In 4.1.0, the EDMC UI appears blurry on my machine:

![blurry UI](https://user-images.githubusercontent.com/759911/95169776-07e7c880-0779-11eb-8217-f5d5443e849e.png)

This is an effect of setting `gdiScaling` to true. If I explicitly override the setting (using the Windows Compatibility properties tab), the UI looks sharp, as it did in 4.0.6:

![sharp UI](https://user-images.githubusercontent.com/759911/95169605-c5be8700-0778-11eb-9fe8-398e603ce672.png)

Note that the GDI-scaled UI does not appear meaningfully larger than the application-scaled UI, despite my system scale being set to 150%.

### To Reproduce

1. Set the system UI scale to 150% in Settings → Display → Scale and layout.
2. Open EDMC.
3. Notice that the UI appears blurry.

### Expected behavior

The UI should be sharp, as it was in 4.0.6.

### Additional context

I’m not an expert on how Windows UI scaling works, but I did a little bit of brief investigation before opening this bug report. [This Microsoft blog post](https://blogs.windows.com/windowsdeveloper/2017/05/19/improving-high-dpi-experience-gdi-based-desktop-apps/) suggests that the issue results from enabling GDI scaling despite the application using text and controls that aren’t actually drawn by GDI. That results in the entire UI being drawn at 100% resolution, then upscaled to 150% using simple bilinear filtering.

The article suggests that GDI scaling may not actually be desirable. Rather, the ideal is for the application to be “DPI aware,” so it draws itself at a higher resolution independently of special operating system intervention. As far as I can tell, EDMC **already did this** in 4.0.6, presumably because whatever GUI library you’re using handled this automatically.

tl;dr: I think enabling `gdiScaling` might actually be more harmful than helpful.

**Comment by Athanasius** [comment]
The problem is that we activated gdiScaling specifically because, when using Windows OS scaling on a resolution higher than 1920x1080 (assuming the physical monitor size didn't compensate and leave the DPI the same) the radio buttons on our 'Appearance' settings tab exhibited weird behaviour. Ref: https://github.com/EDCD/EDMarketConnector/issues/692#issuecomment-691137453

The scaling you get otherwise is at the mercy of what the Tk developers have managed to implement, i.e. we have no control over it.

What screen resolution are you using there ?  It seems weird that you're using EDMC's default theme and yet don't have the smaller text on the menu labels that I saw myself during testing.

Oh and my reading of that article is that in your case everything would be rendered at 200% and then scaled down to 150%.  It "over-renders" and then scales down so as to avoid issues with simply scaling up (which would leave jagged edges).

**Comment by lexi-lambda** [comment]
Thanks for the investigation—what you’ve said makes sense.

This bug is a pretty minor nitpick. 🙂 Obviously I do not spend very much time staring at the EDMC UI, so if it looks a little blurry, I don’t really care. But I figured I’d report the issue, anyway, in case it proved helpful.

**Comment by Athanasius** [comment]
Nope, without gdiScaling set True, and using just the per-display scaling, those radio buttons still end up undersized.
![per-monitor-150pc_no-gdiScaling_radio-buttons-small](https://user-images.githubusercontent.com/118007/95228551-63f13200-07f7-11eb-9098-c1d4571e75bc.png)

**Comment by lexi-lambda** [comment]
Yes, I can confirm that the radio buttons appear undersized for me, too. I think I prefer the undersized radiobuttons with the sharp text to the correctly-scaled radio buttons with blurry text, but clearly neither is ideal.

**Comment by Athanasius** [comment]
I think what we'll do is leave gdiScaling true for now, and see if we can get a small test case to reproduce the radio button issue and report it upstream.

There's also the newer (Tk 8.5 and up) themed widgets to check out, maybe they don't even exhibit this issue.

If we can fix, workaround or avoid the radio button issue (although that's the only bit of core EDMC UI that has them any plugins might have a lot more in their preferences) then we can turn gdiScaling off.

**Comment by lexi-lambda** [comment]
Thanks, that’s certainly fine with me. I appreciate you going to the effort to investigate!

**Comment by FJRMaverick** [comment]
This blurry text thing is not a problem of EDMC alone. Many Windows programs have this blurry text (even ED-Launcher appears quite blurry on 4k and a 60" screen). It maybe a Windows "problem" ....

**Comment by Athanasius** [comment]
> This blurry text thing is not a problem of EDMC alone. Many Windows programs have this blurry text (even ED-Launcher appears quite blurry on 4k and a 60" screen). It maybe a Windows "problem" ....

As I've been playing with a small test program to see if I can report the radio button issue upstream I'd wager those other programs aren't setting dpiAware in their manifest.  *Without* it the radio buttons behave, but the text (per-monitor scaling at 150%) is also fuzzy.  So, activate dpiAware in the .exe manifest to get crisper text, but then you have issues like with these radio buttons.

Then if we activate gdiScaling to fix the radio buttons, that turns off dpiAware and we're back to fuzzy text....

We're seeing issues with the transparent theme and gdiScaling/no dpiAware too, so will probably turn it off in the next release (likely 4.1.3 rather than waiting for a 4.2.0).

**Comment by FJRMaverick** [comment]
mhm, I would regret if gdiScaling is turned off again, because it not only solves the radio buttons issue, but also adopts the icons of the Canon plugin (as an example) and the minimize/close buttons, which were very small before

**Comment by Athanasius** [comment]
The only reason gdiScaling 'fixes' some things is because it causes dpiAware to be ignored, i.e. setting both is equivalent to setting neither in this case.

The choice boils down to:

1. Do not set dpiAware: have fuzzy text when at any scale other than 100%.
1. Set dpiAware: have crisper text, but now radio buttons act up.

I'm putting together a bug report to see if upstream can and will fix this.

**Comment by Athanasius** [comment]
Now I got around to checking for existing issues, this has apparently not only been reported to the Tk devs but also fixed: https://core.tcl-lang.org/tk/info/3c6660b6f0bed337

Now to confirm what version got bundled with Python 3.9.0, given I've reproduced it on that.

**Comment by Athanasius** [comment]
FTR: This was fixed by commits https://github.com/tcltk/tk/commit/bd0d1fd4d2137bebd775874b705c061a7e13aa7f and https://github.com/tcltk/tk/commit/18fd1bf89685a067ea6ad92d6b1d7f260394ddce in https://github.com/tcltk/tk .

**Comment by Athanasius** [comment]
Latest tcl/tk release is Tcl/Tk 8.6.10 (Nov 21, 2019), so it's not too surprising that the Python devs haven't included anything newer than that, even with Python 3.9.0.

So, a future EDMC release will go back to dpiAware true, gdiScaling not set, and we'll just have to wait for the eventual day when 1) Tcl/Tk release a new version with the fix (which was checked in May 2020), and 2) a new Python version that uses that.

**Comment by Athanasius** [comment]
Ref: https://bugs.python.org/issue41969

**Comment by Athanasius** [comment]
I've now taken gdiScaling out of the manifest file in all the relevant branches, so this will be fixed in whatever our next release is.  I'm not yet rushing to make one, but will in at most a month's time if nothing else has come up.

**Comment by Athanasius** [comment]
As 4.1.3 is out with gdiScaling reverted to false (and thus dpiAware now active again), and it calls out the radio button issue in its changelog, I'm closing this.
